packages/react/src/components/Dialog/index.tsx

```diff
@@ -28,196 +28,159 @@ export interface DialogProps extends React.HTMLAttributes<HTMLDivElement> {
-  static defaultProps = {
-    onClose: noop,
-    forceAction: false,
-    closeButtonText: 'Close'
-  };
-
-  private element: HTMLDivElement | null;
-  private heading: HTMLHeadingElement | null;
-  private headingId: string = nextId('dialog-title-');
-
-  constructor(props: DialogProps) {
-    super(props);
-
-    this.close = this.close.bind(this);
-    this.focusHeading = this.focusHeading.bind(this);
-    this.handleClickOutside = this.handleClickOutside.bind(this);
-    this.handleEscape = this.handleEscape.bind(this);
-    this.state = {};
-  }
-
-  componentDidMount() {
-    if (this.props.show) {
-      this.attachEventListeners();
-      this.attachIsolator(() => setTimeout(this.focusHeading));
-    }
-  }
-
-  componentWillUnmount() {
-    const { isolator } = this.state;
-    isolator?.deactivate();
-    this.removeEventListeners();
-  }
-
-  componentDidUpdate(prevProps: DialogProps) {
-    if (!prevProps.show && this.props.show) {
-      this.attachIsolator(this.focusHeading);
-      this.attachEventListeners();
-    } else if (prevProps.show && !this.props.show) {
-      this.removeEventListeners();
-      this.close();
-    }
-  }
-
-  private attachIsolator(done: () => void) {
-    this.setState(
-      {
-        isolator: new AriaIsolate(this.element as HTMLElement)
-      },
-      done
-    );
-  }
-
-  render() {
-    const {
-      dialogRef,
-      forceAction,
+const Dialog = forwardRef<HTMLDivElement, DialogProps>(
+  (
+    {
+      dialogRef: dialogRefProp,
+      forceAction = false,
       className,
       children,
-      closeButtonText,
+      closeButtonText = 'Close',
       heading,
-      show,
+      show = false,
+      portal,
+      onClose = () => null,
       ...other
-    } = this.props;
+    },
+    ref
+  ): React.ReactPortal | null => {
+    const dialogRef = useSharedRef(dialogRefProp || ref);
+    const [headingId] = useId(1, 'dialog-title-');
+    const headingRef = useRef<HTMLHeadingElement>(null);
+    const isolatorRef = useRef<AriaIsolate>();
+
+    const handleClose = useCallback(() => {
+      isolatorRef.current?.deactivate();
+      if (show) {
+        onClose();
+      }
+    }, [show, onClose]);
+
+    const handleClickOutside = useCallback(() => {
+      if (show && !forceAction) {
+        handleClose();
+      }
+    }, [show, forceAction, handleClose]);
+
+    const focusHeading = useCallback(() => {
+      if (headingRef.current) {
+        headingRef.current.focus();
+      }
+      isolatorRef.current?.activate();
+    }, []);
+
+    const handleEscape = useCallback(
+      (keyboardEvent: KeyboardEvent) => {
+        if (!keyboardEvent.defaultPrevented && isEscape(keyboardEvent)) {
+          handleClose();
+        }
+      },
+      [handleClose]
+    );
+
+    useEffect(() => {
+      if (!show || !dialogRef.current) return;
+
+      isolatorRef.current = new AriaIsolate(dialogRef.current);
+      setTimeout(focusHeading);
```

Comment:
nit - this use of `setTimeout` could probably benefit from a comment (I assume this is to wait for the dialog to be rendered, if so there may be a way to avoid setTimeout use)

Reply:
I agree, but I wish I remembered why as it's been like this since it's initial implementation: https://github.com/dequelabs/cauldron/blob/eb5def134fcae7ec53e00aa6a47d2837c6af2783/packages/react/src/components/Dialog/index.tsx#L65-L66

I'm taking an educated guess that it's queuing the focus to be on the next tick to prevent focus conflicts. We can add that as a comment if you feel that's useful?